### PR TITLE
Forbid mute hooks

### PR DESCRIPTION
**Pull request by ayalash**

### CODE REVIEW

File: doc/advanced.rst [hunk 1]
```diff
@@ -192,6 +192,18 @@ You can selectively mute hooks (prevent their callbacks from being called) throu
 		>>> with gossip.mute_context(['my.hook.name']):
 		...     function_that_triggers_hooks()  # <--- nothing happens
 
+However, both hooks and groups can forbid the usage of :func:`.mute_context` on them:
+
+.. code-block:: python
+
+		>>> hook = gossip.define('my_unmuted_hook')
+		>>> hook.should_not_be_muted()
```
Comment (vmalloc) [inline doc/advanced.rst]: @ayalash I think this should be called `forbid_muting` or `disable_muting`

File: doc/advanced.rst [hunk 2]
```diff
@@ -192,6 +192,18 @@ You can selectively mute hooks (prevent their callbacks from being called) throu
...
+		>>> with gossip.mute_context(['my_unmuted_hook']):
+		...     pass
+		Traceback (most recent call last):
+		   ...
+		CannotMuteHooks: Hooks cannot be muted: my_unmuted_hook
```
Comment (vmalloc) [inline doc/advanced.rst]: `... Muting is forbidden for the hook 'my_unmuted_hook'`

File: gossip/groups.py [hunk 3]
```diff
@@ -71,6 +72,16 @@ def set_strict(self, strict=True):
                 child.validate_strict()
         self._strict = strict
 
+    def should_not_be_muted(self):
```
Comment (vmalloc) [inline gossip/groups.py]: See rename note above

File: gossip/groups.py [hunk 4]
```diff
@@ -71,6 +72,16 @@ def set_strict(self, strict=True):
...
+        self._can_be_muted = False
+
+    def can_be_muted(self):
+        if self._can_be_muted is not None:
+            return self._can_be_muted
```
Comment (vmalloc) [inline gossip/groups.py]: Should it be allowed to allow muting on a specific hook of a group which cannot be muted? Not sure...

File: gossip/groups.py [hunk 5]
```diff
@@ -88,12 +99,6 @@ def add_child(self, name, child):
     def remove_child(self, child):
         self._children.pop(child)
 
-    def get_or_create_subgroup(self, name):
```
Comment (vmalloc) [inline gossip/groups.py]: Why was this removed?

File: gossip/hooks.py [hunk 6]
```diff
@@ -35,9 +36,18 @@ def __init__(self, group, name, arg_names=(), doc=None, deprecated=False):
         self._defined = False
         self._pre_trigger_callbacks = []
         self._unmet_deps = frozenset()
+        self._can_be_muted = can_be_muted
         self.doc = doc
         self.deprecated = deprecated
 
+    def should_not_be_muted(self):
```
Comment (vmalloc) [inline gossip/hooks.py]: See rename note above

File: gossip/hooks.py [hunk 7]
```diff
@@ -35,9 +36,18 @@ def __init__(self, group, name, arg_names=(), doc=None, deprecated=False):
...
+        self._can_be_muted = False
+
+    def can_be_muted(self):
```
Comment (vmalloc) [inline gossip/hooks.py]: See https://github.com/getslash/gossip/pull/18/files#r70689639 - similar note here

File: tests/test_gossip.py [hunk 8]
```diff
@@ -180,3 +180,45 @@ def test_mute_accepts_only_lists(arg):
...
+    hook = gossip.define('a.b.d')
+    hook.should_not_be_muted()
+    assert not hook.can_be_muted()
+
+    with gossip.mute_context(['a.b.c']):
+        gossip.trigger('a.b.c')
+    assert not checkpoint.called
+
+
+def mute_other_sub_group():
```
Comment (vmalloc) [inline tests/test_gossip.py]: more tests should be added for muting subgroups inside unmuted groups, hooks and vice-versa (whether we forbid it or not)

File: gossip/groups.py (hunk 5 above)
Comment (ayalash) [inline gossip/groups.py]: This method was broken - it used a private members that weren't exist.

Comment (vmalloc) [inline gossip/groups.py]: Whoops! ok.

**Comment by ayalash** [comment]
@vmalloc This pull request is out of date.
I'm opening a new (updated) pull requests instead: 
* pylint: https://github.com/getslash/gossip/pull/20
* forbid_mute_hooks: TBD
